### PR TITLE
feat(core): add bip32path.fromLegacyPath()

### PR DESCRIPTION
**Pull request by OttoAllmendinger**
Add helper for converting buggy invalid bip32 paths into somewhat
reasonable bip32 paths that can be used with standard libraries.

Issue: BG-34708

### CODE REVIEW

File: modules/core/test/unit/bip32path.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      throw new Error(`invalid state`);
+    }
+
+    if (inputs.length === 0) {
+      throw new Error(`invalid argument`);
+    }
+
+    const refPath = inputs[0];
+
+    fromLegacyPath(refPath).should.eql(refPath);
```
Comment (sangaman) [inline modules/core/test/unit/bip32path.ts:41]: This comparison check is happening a couple lines below as part of the `forEach` loop I believe - I figure this line can be removed so it only happens once?

File: modules/core/test/unit/bip32path.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    inputs.forEach((p) => {
+      // console.log(`legacyPath=${p} path=${refPath}`);
```
Comment (sangaman) [inline modules/core/test/unit/bip32path.ts:44]: Remove commented out code?

**Review by sangaman — APPROVED** [review]
Very nice, seems like a complex issue to deal with but it looks like this test suite should go a long way towards clearing it up. I figure it can be expanded later if other derivation path peculiarities arise.

**Review by tylerlevine — APPROVED** [review]
minor nits but this is blocked on the new audit problems, I will try to take a look later today but this will likely need a rebase

File: modules/core/src/bip32path.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    .map((p) => {
+      const hardened = p.slice(-1) === "'";
+      const v = parseInt(p.slice(0, hardened ? -1 : undefined), 10);
+      if (Number.isNaN(v)) {
+        return '0';
+      }
+      return String(v) + (hardened ? "'" : '');
+    });
+
+  parts.forEach((p, i) => {});
```
[the remark this answers is not in the document]
Comment (OttoAllmendinger) [reply modules/core/src/bip32path.ts:38]: oops

File: modules/core/test/unit/bip32path.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+      fromLegacyPath(p).should.eql(refPath);
+      refBIP32
+        .derivePath(refPath)
+        .toBase58() //
```
[the remark this answers is not in the document]
Comment (OttoAllmendinger) [reply modules/core/test/unit/bip32path.ts:48]: that one is in fact to force a linebreak (for subjective style preference)

I'll get rid of it

File: modules/core/src/bip32path.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+ * This probably does not cover everything but it should cover most bugs encountered in our code.
+ *
+ * This method attempts to convert a "legacy path", which may be invalid, to a standard bip32 path:
+ * meaning the result should be equivalent to the input path when passed to either our custom
+ * derivation log or when passed to a standard bip32 library.
+ *
+ * @param path - legacy path
+ * @return string - somewhat sanitized path that can be passed to standard bip32 libraries
+ */
+export function fromLegacyPath(path: string): string {
```
[the remark this answers is not in the document]
Comment (OttoAllmendinger) [reply modules/core/src/bip32path.ts:24]: `sanitizeLegacyPath()` it is

File: modules/core/test/unit/bip32path.ts (hunk 1 above)
Comment (OttoAllmendinger) [reply modules/core/test/unit/bip32path.ts:41]: the stack trace for this error is slightly clearer, but I should instead improve the error on the other assertion

**Review by tylerlevine — APPROVED** [review]

**Pull request merged**